smart/model_assembly.py

```diff
@@ -1787,3 +1812,43 @@ def sbmodel_from_locals(local_values):
     cc.add(compartments)
     rc.add(reactions)
     return pc, sc, cc, rc
+
+
+def create_restriction(u, mesh_function, value):
```

Comment:
This is essentially a repeat of the `create_restriction` function in `smart.model`, but it needs to be called in `model_assembly` to restrict reactions to a certain subdomain (see definitions of the flux forms)

Reply:
Couldn't we just keep just this function, and change the above to:
```python
    def create_restriction(self, u, mesh_function, value):
        """
        Restrict a function on a submesh to a subset of parent entities
        (same dimension as the submesh)
        :param u: Function on submesh
        :param mesh_function: MeshFunction marking the subset of parent entities
                            (same dimension as the cells of the submesh)
        :param value: Value in MeshFunction marking the subset of parent entities
        :return: New restricted function
        """
        from smart.model_assembly import create_restriction as _c_r
        return _c_r(u, mesh_function_value) 
```
to avoid code duplication.
It would also be good to use type hints here, i.e. `u:dolfin.Function, mesh_function: dolfin.MeshFunction, value: np.integer`